agent/engine/docker_image_manager.go

```diff
@@ -214,21 +216,42 @@ func (imageManager *dockerImageManager) removeImageState(imageStateToBeRemoved *
 }
 
 func (imageManager *dockerImageManager) getCandidateImagesForDeletion() []*image.ImageState {
-	if len(imageManager.imageStatesConsideredForDeletion) < 1 {
+	return imageManager.getCandidateImagesForDeletionHelper(imageManager.imageStatesConsideredForDeletion)
+}
+
+func (imageManager *dockerImageManager) getCandidateImagesForDeletionHelper(candidateImagesConsiderForDeletion map[string]*image.ImageState) []*image.ImageState {
+	if len(candidateImagesConsiderForDeletion) < 1 {
 		seelog.Debugf("Image Manager: Empty state!")
 		// no image states present in image manager
 		return nil
 	}
 	var imagesForDeletion []*image.ImageState
-	for _, imageState := range imageManager.imageStatesConsideredForDeletion {
+	for _, imageState := range candidateImagesConsiderForDeletion {
 		if imageManager.isImageOldEnough(imageState) && imageState.HasNoAssociatedContainers() {
 			seelog.Infof("Candidate image for deletion: [%s]", imageState.String())
+			// remove the images that cannot be added
+			if imageManager.includesImageCleanupExclusionList(imageState) {
+				//imageState that we want to keep
+				seelog.Infof("image excluded from deletion: [%s]", imageState.String())
+				continue
+			}
 			imagesForDeletion = append(imagesForDeletion, imageState)
 		}
 	}
 	return imagesForDeletion
 }
 
+func (ImageManager *dockerImageManager) includesImageCleanupExclusionList(imageState *image.ImageState) bool {
+	for _, imageNotDelete := range ImageManager.imageCleanupExclusionList {
```

Comment:
Does it make more sense if you use a set (or create a set and use it everywhere) for `imageCleanupExclusionList` instead of a list? That way you just need to do one for loop here